luigi/rpc.py

```diff
@@ -42,32 +57,62 @@ def __init__(self, message, sub_exception=None):
         self.sub_exception = sub_exception
 
 
+class FetcherException(Exception):
+    def __init__(self, original_exc):
+        self.original_exc = original_exc
+
+
+class URLLibFetcher(object):
+    def fetch(self, full_url, body, timeout):
+        try:
+            body = urlencode(body).encode('utf-8')
+            return urlopen(full_url, body, timeout).read().decode('utf-8')
+        except (URLError, socket.timeout) as e:
+            raise FetcherException(e)
+
+
+class RequestsFetcher(object):
+    def __init__(self, session):
+        self.session = session
+
+    def fetch(self, full_url, body, timeout):
+        from requests import exceptions as requests_exceptions
+        try:
+            resp = self.session.get(full_url, data=body, timeout=timeout)
+            resp.raise_for_status()
+            return resp.text
+        except (requests_exceptions.RequestException) as e:
+            raise FetcherException(e)
+
+
 class RemoteScheduler(Scheduler):
     """
     Scheduler proxy object. Talks to a RemoteSchedulerResponder.
     """
 
-    def __init__(self, host='localhost', port=8082, connect_timeout=None, url_prefix=''):
-        self._host = host
-        self._port = port
-        self._url_prefix = url_prefix
+    def __init__(self, url='http://localhost:8082/', connect_timeout=None):
+        assert (
+            not (url.startswith('http+unix://') and not HAS_UNIX_SOCKETS),
+            'You need to install requests-unixsocket for Unix socket support.',
+        )
 
+        self._url = url
         config = configuration.get_config()
 
         if connect_timeout is None:
             connect_timeout = config.getfloat('core', 'rpc-connect-timeout', 10.0)
         self._connect_timeout = connect_timeout
 
+        if HAS_REQUESTS:
+            self._fetcher = RequestsFetcher(requests.Session())
+        else:
+            self._fetcher = URLLibFetcher()
+
     def _wait(self):
         time.sleep(30)
 
     def _fetch(self, url, body, log_exceptions=True, attempts=3):
-
-        full_url = 'http://{host}:{port:d}{prefix}{url}'.format(
-            host=self._host,
-            port=self._port,
-            prefix=self._url_prefix,
-            url=url)
+        full_url = self._url + url
```

Comment:
Erm, there's no way this line can be "right". Are the variable names extremely confusing? Or you actually string-concatenating two urls?